src/librustc_typeck/check/mod.rs

```diff
@@ -1322,19 +1322,41 @@ fn check_union(tcx: TyCtxt<'_>, id: hir::HirId, span: Span) {
     check_packed(tcx, span, def_id);
 }
 
-fn check_opaque<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, substs: SubstsRef<'tcx>, span: Span) {
+fn check_opaque<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    substs: SubstsRef<'tcx>,
+    span: Span,
+    origin: &hir::ExistTyOrigin
+) {
     if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id, substs) {
-        let mut err = struct_span_err!(
-            tcx.sess, span, E0720,
-            "opaque type expands to a recursive type",
-        );
-        err.span_label(span, "expands to a recursive type");
-        if let ty::Opaque(..) = partially_expanded_type.sty {
-            err.note("type resolves to itself");
-        } else {
-            err.note(&format!("expanded type is `{}`", partially_expanded_type));
+        match origin {
+            hir::ExistTyOrigin::AsyncFn =>{
+                let mut err = struct_span_err!(
+                    tcx.sess, span, E0733,
+                    "recursion in an async fn requires boxing",
+                );
+
+                err.span_label(span, "async fn cannot invoke themselves directly");
+                err.note(r##"to create a recursive async fn, it must be rewritten to return a
+boxed future.
+For more information, see https://rust-lang.github.io/async-book/"##);
```

Comment:
Is there a more targeted link available? This is throwing the user into the "middle of nowhere".

Reply:
I'd just leave this out entirely-- anything pointing into that book is subject to change in the near future.